File: src/main/scala/edu/umass/cs/iesl/nndepparse/ParseState.scala
package edu.umass.cs.iesl.nndepparse

import scala.annotation.tailrec
import scala.collection.mutable.Set

/* Represents current stack, parse, input during transition-based parsing, super mutable */
class ParseState(var stack: Int, var input: Int, val reducedIds: Set[Int], val sentence: LightweightParseSentence) {
  val parseSentenceLength = sentence.length

  val headIndices = Array.fill[Int](parseSentenceLength)(-1)
  val arcLabels = Array.fill[String](parseSentenceLength)("")

  val leftmostDeps = Array.fill[Int](parseSentenceLength)(-1)
  val rightmostDeps = Array.fill[Int](parseSentenceLength)(-1)

  val leftmostDeps2 = Array.fill[Int](parseSentenceLength)(-1)
  val rightmostDeps2 = Array.fill[Int](parseSentenceLength)(-1)

  def goldHeads = sentence.goldHeads
  def goldLabels = sentence.goldLabels

  def setHead(tokenIndex: Int, headIndex: Int, label: String) = {
    // set head
    headIndices(tokenIndex) = headIndex
    arcLabels(tokenIndex) = label

    // update left and rightmost dependents
    if(headIndex != -1){
      if (tokenIndex < leftmostDeps(headIndex) || leftmostDeps(headIndex) == -1) {
        leftmostDeps2(headIndex) = leftmostDeps(headIndex)
        leftmostDeps(headIndex) = tokenIndex
      }
      if(tokenIndex > rightmostDeps(headIndex) || rightmostDeps(headIndex) == -1) {
        rightmostDeps2(headIndex) = rightmostDeps(headIndex)
        rightmostDeps(headIndex) = tokenIndex
      }
    }
  }

  @tailrec final def isDescendantOf(firstIndex: Int, secondIndex: Int): Boolean = {
    val firstHeadIndex = headIndices(firstIndex)
    if (firstHeadIndex == -1) false // firstIndex has no head, so it can't be a descendant
    else if (firstHeadIndex == secondIndex) true
    else isDescendantOf(firstHeadIndex, secondIndex)
  }

  def leftmostDependent(tokenIndex: Int): Int = {
    if (tokenIndex == -1) -1
    else leftmostDeps(tokenIndex)
  }

  def rightmostDependent(tokenIndex: Int): Int = {
    if (tokenIndex == -1) -1
    else rightmostDeps(tokenIndex)
  }

  def leftmostDependent2(tokenIndex: Int): Int = {
    if (tokenIndex == -1) -1
    else leftmostDeps2(tokenIndex)
  }

  def rightmostDependent2(tokenIndex: Int): Int = {
    if (tokenIndex == -1) -1
    else rightmostDeps2(tokenIndex)
  }

  def grandLeftmostDependent(tokenIndex: Int): Int = {
    if (tokenIndex == -1) -1
    else{
      val i = leftmostDeps(tokenIndex)
      if (i == -1) -1
      else leftmostDeps(i)
    }
  }

  def grandRightmostDependent(tokenIndex: Int): Int = {
    if (tokenIndex == -1) -1
    else {
      val i = rightmostDeps(tokenIndex)
      if (i == -1) -1
      else rightmostDeps(i)
    }
  }

  def leftNearestSibling(tokenIndex: Int): Int = {
    val tokenHeadIndex = headIndices(tokenIndex)
    if(tokenHeadIndex != -1){
      var i = tokenIndex - 1
      while(i >= 0){
        if (headIndices(i) != -1 && headIndices(i) == tokenHeadIndex)
          return i
        i -= 1
      }
    }
    -1
  }

  def rightNearestSibling(tokenIndex: Int): Int = {
    val tokenHeadIndex = headIndices(tokenIndex)
    if(tokenHeadIndex != -1){
      var i = tokenIndex + 1
      while(i < parseSentenceLength){
        if(headIndices(i) != -1 && headIndices(i) == tokenHeadIndex)
          return i
        i += 1
      }
    }
    -1
  }

  def inputToken(offset: Int): Int = {
    val i = input + offset
    if (i < 0 || parseSentenceLength - 1 < i) -1
    else i
  }

  def lambdaToken(offset: Int): Int = {
    val i = stack + offset
    if (i < 0 || parseSentenceLength - 1 < i) -1
    else i
  }

  def stackToken(offset: Int): Int = {
    if (offset == 0)
      return stack
    var off = math.abs(offset)
    var dir = if (offset < 0) -1 else 1
    var i = stack + dir
    while (0 < i && i < input) {
      if (!reducedIds.contains(i)) {
        off -= 1
        if (off == 0)
          return i
      }
      i += dir
    }
    i
  }

  def print() = {
    println(s"heads: ${headIndices.mkString(" ")}")
    println(s"labels: ${arcLabels.map(l => if(l == "") "-1" else l).mkString(" ")}")
    println(s"gold heads: ${goldHeads.mkString(" ")}")
    println(s"gold labels: ${goldLabels.mkString(" ")}")
    println(s"gold pos: ${(0 until parseSentenceLength).map(i => sentence(i).posTagString).mkString(" ")}")
    println(s"stack: $stack stack-1: ${stackToken(-1)} input: $input")
    println(s"reduced ids: ${reducedIds.mkString(" ")}")
  }
}

File: src/main/lua/ParseState.lua
require 'torch'
require 'try-catch'

package.path = package.path .. ";src/main/lua/?.lua"

local ParseState = torch.class('ParseState')

-- constructor
function ParseState:__init(stack, input, reducedIds, sentence, pos_features)
    self.__index = self
    self.parseSentenceLength = sentence:size(1) + 1
    self.stack = stack
    self.input = input
    self.reducedIds = reducedIds
    self._sentence = sentence
    self.headIndices = torch.Tensor(self.parseSentenceLength):fill(0)
    self.arcLabels = torch.Tensor(self.parseSentenceLength):fill(0)

    self.leftmostDeps = torch.Tensor(self.parseSentenceLength):fill(0)
    self.rightmostDeps = torch.Tensor(self.parseSentenceLength):fill(0)
    self.leftmostDeps2 = torch.Tensor(self.parseSentenceLength):fill(0)
    self.rightmostDeps2 = torch.Tensor(self.parseSentenceLength):fill(0)

    -- todo should maybe actually be a root representation
    self.features = pos_features

end

function ParseState:print()
    print("sentence length: ", self.parseSentenceLength)
    print(string.format("stack: %d stack-1: %d input: %d", self.stack, self:stackToken(-1), self.input))
    print(string.format("reduced ids: %s", self:mk_string(self.reducedIds, " ")))
    print("arc labels: ")
    print(self:mk_string_tensor(self.arcLabels, " "))
    print("head indices: ")
    print(self:mk_string_tensor(self.headIndices, " "))

    print("true arc labels:")
    print(self:mk_string_tensor(self._sentence:select(2,3), " "))
    print("true heads:")
    print(self:mk_string_tensor(torch.add(self._sentence:select(2,4), 2), " "))


    print("leftmost deps: ")
    print(self:mk_string_tensor(self.leftmostDeps, " "))
    print("rightmost deps: ")
    print(self:mk_string_tensor(self.rightmostDeps, " "))
end

function ParseState:mk_string(t, sep)
    local str = ""
    if(t) then
        for i,_ in pairs(t) do
            str = str .. i .. sep
        end
    end
    return str
end

function ParseState:mk_string_tensor(tensor, sep)
    local str = ""
    if(tensor) then
        for i = 1,tensor:size()[1] do
            str = str .. tensor[i] .. sep
        end
    end
    return str
end

function ParseState:sentence(index)
    if(index == 1) then
        return Constants.ROOT[1]
    elseif (index > 1 and index <= self.parseSentenceLength) then
        return self._sentence[index-1]
    else
        return Constants.NULL[1]
    end
end

function ParseState:setHead(tokenIndex, headIndex, label)

        self.headIndices[tokenIndex] = headIndex
        self.arcLabels[tokenIndex] = label

        if(headIndex > 0) then
            if(tokenIndex < self.leftmostDeps[headIndex] or self.leftmostDeps[headIndex] == 0) then
                self.leftmostDeps2[headIndex] = self.leftmostDeps[headIndex]
                self.leftmostDeps[headIndex] = tokenIndex
            end
            if(tokenIndex > self.rightmostDeps[headIndex] or self.rightmostDeps[headIndex] == 0) then
                self.rightmostDeps2[headIndex] = self.rightmostDeps[headIndex]
                self.rightmostDeps[headIndex] = tokenIndex
            end
        end
--    end
end

function ParseState:isDescendantOf(firstIndex, secondIndex)
    local firstHeadIndex = self.headIndices[firstIndex]
    if(firstHeadIndex == 0) then return false -- firstIndex has no head, so it can't be a descendant
    elseif(firstHeadIndex == secondIndex) then return true
    else return self:isDescendantOf(firstHeadIndex, secondIndex)
    end
end

function ParseState:leftmostDependent(tokenIndex)
    if(tokenIndex <= 0 or tokenIndex >= self.parseSentenceLength) then return 0
    else return self.leftmostDeps[tokenIndex]
    end
end

function ParseState:rightmostDependent(tokenIndex)
    if(tokenIndex <= 0 or tokenIndex >= self.parseSentenceLength) then return 0
    else return self.rightmostDeps[tokenIndex]
    end
end

function ParseState:leftmostDependent2(tokenIndex)
    if(tokenIndex <= 0 or tokenIndex >= self.parseSentenceLength) then return 0
    else return self.leftmostDeps2[tokenIndex]
    end
end

function ParseState:rightmostDependent2(tokenIndex)
    if(tokenIndex <= 0 or tokenIndex >= self.parseSentenceLength) then return 0
    else return self.rightmostDeps2[tokenIndex]
    end
end

function ParseState:grandLeftmostDependent(tokenIndex)
    if(tokenIndex <= 0 or tokenIndex >= self.parseSentenceLength) then return 0
    else
        local i = self.leftmostDeps[tokenIndex]
        if(i == 0) then return 0
        else return self.leftmostDeps[i]
        end
    end
end

function ParseState:grandRightmostDependent(tokenIndex)
    if(tokenIndex <= 0 or tokenIndex >= self.parseSentenceLength) then return 0
    else
        local i = self.rightmostDeps[tokenIndex]
        if(i == 0) then return 0
        else return self.rightmostDeps[i]
        end
    end
end

function ParseState:leftNearestSibling(tokenIndex)
    local tokenHeadIndex = self.headIndices[tokenIndex]
    if(tokenHeadIndex ~= 0) then
        local i = tokenIndex - 1
        while(i > 0) do
            if(self.headIndices[i] ~= 0 and self.headIndices[i] == tokenHeadIndex) then
                return i
            end
            i = i - 1
        end
    end
    return 0
end

function ParseState:rightNearestSibling(tokenIndex)
    local tokenHeadIndex = self.headIndices[tokenIndex]
    if(tokenHeadIndex ~= 0) then
        local i = tokenIndex + 1
        while(i <= self.parseSentenceLength) do
            if(self.headIndices[i] ~= 0 and self.headIndices[i] == tokenHeadIndex) then
                return i
            end
            i = i + 1
        end
    end
    return 0
end

function ParseState:inputToken(offset)
    local i = self.input + offset
    if(i < 1 or self.parseSentenceLength < i) then return 0
    else return i
    end
end

function ParseState:lambdaToken(offset)
    local i = self.stack + offset
    if(i < 1 or self.parseSentenceLength < i) then return 0
    else return i
    end
end

function ParseState:stackToken(offset)
    if(offset == 0) then return self.stack end
    local off = math.abs(offset)
    local dir = 1
    if(offset < 0) then dir = -1 end
    local i = self.stack + dir
    while(1 < i and i < self.input) do
        if(not self.reducedIds[i]) then
            off = off - 1
            if(off == 0) then return i end
        end
        i = i + dir
    end
    return i
end